pybop/observers/unscented_kalman.py

```diff
@@ -0,0 +1,309 @@
+        return (points, w_m, w_c)
+
+    @staticmethod
+    def unscented_transform(
+        sigma_points: np.ndarray, w_m: np.ndarray, w_c: np.ndarray, sqrtR: np.ndarray
+    ) -> Tuple[np.ndarray, np.ndarray]:
+        """
+        Performs the unscented transform
+
+        Parameters
+        ----------
+        sigma_points : List[SigmaPoint]
+            The sigma points
+        sqrtR : np.ndarray
+            The square root of the covariance matrix
+
+        Returns
+        -------
+        Tuple[np.ndarray, np.ndarray]
+            The mean and covariance of the sigma points
+        """
+        x = np.sum(w_m * sigma_points, axis=1).reshape(-1, 1)
+        A = np.sqrt(w_c[1:]) * (sigma_points[:, 1:] - x)
+        (
+            _,
+            S,
+        ) = linalg.qr(A.T, mode="economic")
+        S = UkfFilter.cholupdate(S, sigma_points[:, 0:1] - x, w_c[0])
+        return x, S
+
+    @staticmethod
+    def cholupdate(R: np.ndarray, x: np.ndarray, w: float) -> np.ndarray:
+        """
+        Updates the cholesky decomposition of a matrix (see https://github.com/modusdatascience/choldate/blob/master/choldate/_choldate.pyx)
+
+        Note: will be in scipy soon so replace with this: https://github.com/scipy/scipy/pull/16499
+
+        TODO: need to replace with something more low-level
+
+        Parameters
+        ----------
+        R : np.ndarray
+            The cholesky decomposition of the matrix
+        x : np.ndarray
+            The vector to add to the matrix
+        w : float
+            The weight of the vector
+
+        Returns
+        -------
+        np.ndarray
+            The updated cholesky decomposition
+        """
+        x = x.flatten()
+        p = x.shape[0]
+        for k in range(p):
+            r = np.sqrt(R[k, k] ** 2 + w * x[k] ** 2)
+            # r = UkfFilter.hypot(R[k, k], x[k])
+            c = r / R[k, k]
+            s = x[k] / R[k, k]
+            R[k, k] = r
+            if k < p - 1:
+                R[k, k + 1 :] = (R[k, k + 1 :] + w * s * x[k + 1 :]) / c
+                x[k + 1 :] = c * x[k + 1 :] - s * R[k, k + 1 :]
+        return R
+
+    def step(self, y: np.ndarray) -> float:
+        """
+        Steps the filter forward one step using a measurement. Returns the log likelihood of the measurement.
+
+        Parameters
+        ----------
+        y : np.ndarray
+            The measurement vector
+
+        Returns
+        -------
+        float
+            The log likelihood of the measurement
+        """
+        sigma_points, w_m, w_c = self.gen_sigma_points(
+            self.x, self.S, self.alpha, self.beta
+        )
+        sigma_points = np.apply_along_axis(self.f, 0, sigma_points)
+
+        x_minus, S_minus = self.unscented_transform(sigma_points, w_m, w_c, self.sqrtRp)
+        sigma_points_y = np.apply_along_axis(self.h, 0, sigma_points)
+        y_minus, S_y = self.unscented_transform(sigma_points_y, w_m, w_c, self.sqrtRm)
+        P = np.sum(
```

Comment:
`np.einsum` might help with performance here. Something like:

```
P = np.einsum('i,ijk,ijl->kl', w_c, sigma_points - x_minus, sigma_points_y - y_minus)
```

Reply:
I'd not used `einsum` before, looks good! I'll change this

Reply:
@BradyPlanden: I tried and failed at this :( the string `i, ijk, ijl` does not work because sigma_points - x_minus is only a 2D array, not 3D. I couldn't come up with the right string to first do the outer product, and then the sub over axes 1 + 3. Any ideas?

Reply:
actually this seems to work with the sizes, but gives a different answer than the previous implementation!

```python
P = np.einsum("k,jk,lk -> jl ", w_c, sigma_points - x_minus, sigma_points_y - y_minus)
```

Reply:
tests still pass though! I think I need a better test on the covariance calculation.....

Reply:
actually the new einsum implementation is the correct answer :) fixed now

Reply:
still needs some better testing. Perhaps after this is merged in I can implement a kalman filter as well. For a linear model the kalman filter and the ukf should give identical answers correct?